File: backend/internal/vehicles/vin_test.go
package vehicles

import (
	"errors"
	"testing"
)

func TestIsValidVIN(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name string
		vin  string
		want bool
	}{
		{
			name: "valid VIN with numeric check digit",
			vin:  "1HGCM82633A004352",
			want: true,
		},
		{
			name: "valid VIN with X check digit",
			vin:  "1M8GDM9AXKP042788",
			want: true,
		},
		{
			name: "valid VIN with lowercase input",
			vin:  "1hgcm82633a004352",
			want: true,
		},
		{
			name: "invalid VIN length",
			vin:  "1HGCM82633A00435",
			want: false,
		},
		{
			name: "invalid VIN checksum",
			vin:  "1HGCM82633A004353",
			want: false,
		},
		{
			name: "invalid VIN contains disallowed character I",
			vin:  "1HGIM82633A004352",
			want: false,
		},
		{
			name: "invalid VIN contains disallowed character O",
			vin:  "1HGOM82633A004352",
			want: false,
		},
		{
			name: "invalid VIN contains disallowed character Q",
			vin:  "1HGQM82633A004352",
			want: false,
		},
	}

	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			got := IsValidVIN(tt.vin)
			if got != tt.want {
				t.Fatalf("IsValidVIN(%q) = %v, want %v", tt.vin, got, tt.want)
			}
		})
	}
}

func TestValidateVINForCreateAndUpdate(t *testing.T) {
	t.Parallel()

	if err := ValidateVINForCreate("1HGCM82633A004352"); err != nil {
		t.Fatalf("ValidateVINForCreate() unexpected error: %v", err)
	}
	if err := ValidateVINForUpdate("1M8GDM9AXKP042788"); err != nil {
		t.Fatalf("ValidateVINForUpdate() unexpected error: %v", err)
	}

	if err := ValidateVINForCreate("INVALID-VIN"); !errors.Is(err, ErrInvalidVIN) {
		t.Fatalf("ValidateVINForCreate() expected ErrInvalidVIN, got: %v", err)
	}
	if err := ValidateVINForUpdate("123"); !errors.Is(err, ErrInvalidVIN) {
		t.Fatalf("ValidateVINForUpdate() expected ErrInvalidVIN, got: %v", err)
	}
}

File: frontend/src/utils/vin.test.ts
import { describe, expect, it } from 'vitest'
import { isValidVin } from './vin'

describe('isValidVin', () => {
  it('accepts 17-character alphanumeric VIN without checksum validation', () => {
    expect(isValidVin('1HGCM82633A004353')).toBe(true)
  })

  it('rejects VIN with invalid length', () => {
    expect(isValidVin('1HGCM82633A00435')).toBe(false)
  })

  it('rejects VIN containing disallowed letters I, O, Q', () => {
    expect(isValidVin('1HGIM82633A004352')).toBe(false)
    expect(isValidVin('1HGOM82633A004352')).toBe(false)
    expect(isValidVin('1HGQM82633A004352')).toBe(false)
  })

  it('rejects VIN containing non-alphanumeric characters', () => {
    expect(isValidVin('1HGCM82633A00-352')).toBe(false)
  })
})
